components/script/dom/paintworkletglobalscope.rs

```diff
@@ -54,44 +84,223 @@ impl PaintWorkletGlobalScope {
+        let _ = sender.send(Ok(image));
+    }
+
+    fn placeholder_image(&self, width: u32, height: u32, pixel: [u8; 4]) -> Image {
         let area = (width as usize) * (height as usize);
         let old_buffer_size = self.buffer.borrow().len();
         let new_buffer_size = area * 4;
-        debug!("Drawing a paint image {}({},{}).", name, width, height);
-        // TODO: call into script to create the image.
-        // For now, we just build a dummy.
         if new_buffer_size > old_buffer_size {
-            let pixel = [0xFF, 0x00, 0x00, 0xFF];
             self.buffer.borrow_mut().extend(pixel.iter().cycle().take(new_buffer_size - old_buffer_size));
         } else {
             self.buffer.borrow_mut().truncate(new_buffer_size);
         }
-        let image = Image {
+        Image {
             width: width,
             height: height,
             format: PixelFormat::RGBA8,
             bytes: IpcSharedMemory::from_bytes(&*self.buffer.borrow()),
             id: None,
-        };
-        let _ = sender.send(Ok(image));
+        }
     }
 }
 
 impl PaintWorkletGlobalScopeMethods for PaintWorkletGlobalScope {
+    #[allow(unsafe_code)]
     /// https://drafts.css-houdini.org/css-paint-api/#dom-paintworkletglobalscope-registerpaint
-    fn RegisterPaint(&self, name: DOMString, _paintCtor: Rc<VoidFunction>) {
+    fn RegisterPaint(&self, name: DOMString, paint_ctor: Rc<VoidFunction>) -> Fallible<()> {
+        let name = Atom::from(name);
+        let cx = self.worklet_global.get_cx();
+        rooted!(in(cx) let paint_obj = paint_ctor.callback_holder().get());
+
         debug!("Registering paint image name {}.", name);
-        // TODO
+
+        // Step 1.
+        if name.is_empty() {
+            return Err(Error::Type(String::from("Empty paint name."))) ;
+        }
+
+        // Step 2-3.
+        if self.paint_definitions.borrow().contains_key(&name) {
+            return Err(Error::InvalidModification);
+        }
+
+        // Step 4-6.
+        debug!("Getting input properties.");
+        let input_properties: Vec<DOMString> =
+            unsafe { get_property(cx, paint_obj.handle(), "inputProperties", StringificationBehavior::Default) }?
+            .unwrap_or_default();
+        debug!("Got {:?}.", input_properties);
+
+        // Step 7-9.
+        debug!("Getting input arguments.");
+        let input_arguments: Vec<DOMString> =
+            unsafe { get_property(cx, paint_obj.handle(), "inputArguments", StringificationBehavior::Default) }?
+            .unwrap_or_default();
+        debug!("Got {:?}.", input_arguments);
+
+        // TODO: Steps 10-11.
+
+        // Steps 12-13.
+        debug!("Getting alpha.");
+        let alpha: bool =
+            unsafe { get_property(cx, paint_obj.handle(), "alpha", ()) }?
+            .unwrap_or_default();
+        debug!("Got {:?}.", alpha);
+
+        // Step 14
+        if unsafe { !IsConstructor(paint_obj.get()) } {
```

Comment:
@asajeffrey `IsConstructor` returns true for any wrapper that is callable which includes wrapped arrow functions. If you use `UnwrapObject` and then use `IsConstructor`, it will return false for the arrow function.